aptos-move/framework/aptos-framework/sources/coin.move

```diff
@@ -348,49 +733,86 @@ module aptos_framework::coin {
         account_addr: address,
         amount: u64,
         burn_cap: &BurnCapability<CoinType>,
-    ) acquires CoinInfo, CoinStore {
+    ) acquires CoinInfo, CoinStore, CoinConversionMap {
         // Skip burning if amount is zero. This shouldn't error out as it's called as part of transaction fee burning.
         if (amount == 0) {
             return
         };
 
-        let coin_store = borrow_global_mut<CoinStore<CoinType>>(account_addr);
-        let coin_to_burn = extract(&mut coin_store.coin, amount);
-        burn(coin_to_burn, burn_cap);
+        let (coin_amount_to_burn, fa_amount_to_burn) = calculate_amount_to_withdraw<CoinType>(
+            account_addr,
+            amount
+        );
+        if (coin_amount_to_burn > 0) {
+            let coin_store = borrow_global_mut<CoinStore<CoinType>>(account_addr);
+            let coin_to_burn = extract(&mut coin_store.coin, coin_amount_to_burn);
+            burn(coin_to_burn, burn_cap);
+        };
+        if (fa_amount_to_burn > 0) {
+            let store_addr = primary_fungible_store::primary_store_address(
+                account_addr,
+                option::destroy_some(paired_metadata<CoinType>())
+            );
+            let fa = fungible_asset::withdraw_internal(store_addr, fa_amount_to_burn);
+            fungible_asset::burn_internal(fa);
```

Comment:
`withdraw_internal` skip the frozen checks, such as `burn_from` itself.
we can use public api.. But I think it will be more costly...let me change it.